_posts/2022-09-15

```diff
@@ -0,0 +1,128 @@
+
+What will `DisposeCounter` be if the resolved type does not implement `IDisposable` (but the implementation does)?
+
+```csharp
+serviceCollection.AddScoped<I1>(_ => new Impl());
+
+// build IServiceProvider
+
+using (var scope = serviceProvider.CreateScope())
+{
+    var service = scope.ServiceProvider.GetService<I1>();
+}
+```
+
+And does `DisposeCounter` change if we resolve the instance from the container too?
+
+```csharp
+serviceCollection.AddScoped<I1>(sp => sp.GetRequiredService<Impl>());
+serviceCollection.AddScoped<Impl>();
+
+// build IServiceProvider
+
+using (var scope = serviceProvider.CreateScope())
+{
+    var service = scope.ServiceProvider.GetService<I1>();
+}
+```
+
+And what happens if we resolve another interface which provides no compile-time information about the resolved implementation?
+
+```csharp
+serviceCollection.AddScoped<I1>(sp => sp.GetRequiredService<I2>());
+serviceCollection.AddScoped<I2>(sp => sp.GetRequiredService<Impl>());
+serviceCollection.AddScoped<Impl>();
+
+// build IServiceProvider
+
+using (var scope = serviceProvider.CreateScope())
+{
+    var service = scope.ServiceProvider.GetService<I1>();
+}
+```
+
+As always, the answer is "it depends". In our case, it depends on the dependency injection container used.
+
+## Microsoft ServiceProvider
+
+Let's start with the Microsoft ServiceProvider. By now it's not just one of the most used DI containers but also pretty much acts as a reference implementation for many other DI containers.
+
+If we build the `IServiceProvider` using `var serviceProvider = serviceCollection.BuildServiceProvider()` we get the following `DisposeCounter` values:
+
+1. 1
+1. 2
+1. 3
+
+From that we can conclude that the implementation [inspects the resolved instances at runtime](https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs#L53) whether they implement `IDisposal` as the container registration can't determine the runtime type of the factory method (or at least not easily). However, The container doesn't deduplicate the disposable references to the same object instance (resolving the target service multiple times won't change the results).
+
+## LightInject
+
+Did you assume (or guess) the correct results for the ServiceProvider? If not, don't worry, there is still a chance a differt dependency injection container behaves the way you expected it to. Let's look at the results when we use LightInject by creating the `IServiceProvider` using `serviceCollection.CreateLightInjectServiceProvider()`:
+
+1. 1
+1. 1
+1. 1
+
+Similar to the Microsoft implementation, LightInject is also able to detect resolved instances implementing `IDisposable`. But LightInject avoids disposing the same instance multiple times.
```

Comment:
```suggestion
Similar to the Microsoft implementation, LightInject can detect resolved instances implementing `IDisposable`. But LightInject avoids disposing the same instance multiple times.
```

~~Is it worth mentioning that the dispose implementation guidelines suggest designing dispose s that it can be called multiple times? As in, don't assume dedup~~ it's mentioned below 😉